File: src/990. Satisfiability of Equality Equations.cpp
class Solution {
public:
    int find(vector<int> &uf, int x){
        if(x != uf[x]) uf[x] = find(uf, uf[x]);
        return uf[x];
    }
    bool equationsPossible(vector<string>& equations) {
        vector<int> uf(26 ,0);
        for(int i = 0; i < 26; ++i) {
            uf[i] = i;
        }
        
        for(auto e : equations){
            if(e[1] == '='){
                int ra = find(uf, e[0] - 'a');
                int rb = find(uf, e[3] - 'a');
                uf[ra] = rb;
            }
        }
        
        for(auto e : equations){
            if(e[1] == '!' && find(uf, e[0] - 'a') == find(uf, e[3] - 'a')) {
                return false;
            }
        }
        return true;
    }
};

File: src/990. Satisfiability of Equality Equations.py
class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        uf = list(range(26))

        def find(x: int) -> int:
            if x != uf[x]:
                uf[x] = find(uf[x])
            return uf[x]

        def getIndex(c: string) -> int:
            return ord(c) - 97
        
        for a, e, _, b in equations:
            if e == "=":
                ra = find(getIndex(a))
                rb = find(getIndex(b))
                uf[ra] = uf[rb]

        return not any(e == "!" and find(getIndex(a)) == find(getIndex(b)) for a, e, _, b in equations)